### ISSUE TITLE
Ruffle hangs on initialization in UXP browsers

### ISSUE THREAD

**Issue by wolfbeast** [ruffle-rs/ruffle#3729]
We're running into a hard hang on initialization in Pale Moon and similar UXP-based browsers. This also occurs on websites that have plugin fallback (since we support NPAPI!) aside Ruffle, and is effectively hanging the only browsers that still natively support the Flash plugin in release versions.
Unfortunately the only message displayed is that the script becomes unresponsive; nothing else is in the consoles.

From your demo page. On loading (after stopping the script by the hang monitor):
```
07:55:10.658 Error: Script terminated by timeout at:
ee@https://ruffle.rs/demo/index.js:1:54845
createPlayer@https://ruffle.rs/demo/index.js:1:83370
h@https://ruffle.rs/demo/index.js:1:84772
@https://ruffle.rs/demo/index.js:1:85734
 1 index.js:1:54845
```

Easily reproducible by downloading a UXP-based browser ([Pale Moon](https://www.palemoon.org) or [Basilisk](https://www.basilisk-browser.org), for example) and loading up the demo page on the ruffle.rs website.

**Comment by Herschel** [comment]
Also reported in #2568. Some investigation is necessary to figure out what exactly is causing the hang (help welcome for those that are familiar with these browsers). It might be useful to sniff these browsers and bail out on execution in the mean time.

**Comment by wolfbeast** [comment]
As said I did have a quick look to see if I coukld garner anything from console messages but it's both minified and quiet code so I couldn't garner anything from it yet.
BTW I'm the main dev behind Pale Moon and UXP. Without more of a hint where to look though, I can't really provide any more information myself as to what goes wrong or what feature to test for to bail out, aside from the obvious of course if NPAPI is supported in the browser natively.

**Comment by Toad06** [comment]
Reproduced on Basilisk.

https://github.com/ruffle-rs/ruffle/blob/760e8421ab2b41ed11e124334d81325c7e0f235e/web/packages/core/src/register-element.ts#L73-L93
The issue is that `window.customElements` is `undefined` in Basilisk, so `window.customElements.define` trigger a `TypeError` and we get stuck in the loop.

**Comment by wolfbeast** [comment]
Yeah we do not have Google WebComponents (CustomElements&shadowDOM) in UXP - at least not yet; we're close but it's an all-or-nothing spec and we might not be able to get the last bits in soon since it's a really complex thing that completely redefines page structure. Waterfox Classic also won't have it.
I'm not sure why you want to rely on CustomElements for Ruffle anyway? I doubt that pages with flash content are going to do webapp-y stuff the Google way.

**Comment by Herschel** [comment]
Hang fixed thanks to #3796. On browsers without custom elements, the Ruffle polyfill should error out early and allow the Flash NPAPI plugin to function normally, if available. If a site is using the Ruffle JS API, Ruffle will error out and no player will be displayed, as there is not a Flash embed in this case; I consider it up up to the website's discretion if they want to provide the option for a standard Flash embed (see Newgrounds for an example).

There was some discussion about abandoning at least custom elements for other reasons, so we might pursue that in the future. I think we'd still want the shadow DOM regardless.

**Comment by danielhjacobs** [comment]
Worth noting, the current error on Pale Moon and Basilisk seems to be `wasm validation error: at offset 78890: unrecognized opcode: c0` (offset may vary). This may be due to incomplete WASM support on UXP browsers.
